### PR TITLE
feat: add failure reporting to TaskTracker

### PR DESCRIPTION
**Pull request by tustvold**
Currently the TaskTracker system (which then feeds the longrunning operations system) only reports when all tracked futures have completed, not if they actually completed successfully. This is somewhat unfortunate.

This PR changes TaskTracker to track three different types of failure:

* User initiated cancellation
* Drop without completion (e.g. task panic-ed)
* Completion with error

The rest of the PR is then an extended plumbing exercise. A consequence of this change is you can now only track a TryFuture, one that returns a Result. It is fairly easy to convert an infallible future to a fallible one, and this avoids having infallible and fallible TrackedFuture

### CODE REVIEW

File: tracker/src/task/future.rs [hunk 1]
```diff
@@ -34,48 +34,64 @@ pub trait TrackedFutureExt: Future {
...
+    complete: bool,
 }
 
-impl<F: Future> Future for TrackedFuture<F> {
-    type Output = Result<F::Output, future::Aborted>;
+impl<F: TryFuture> Future for TrackedFuture<F> {
+    type Output = Result<Result<F::Ok, F::Error>, future::Aborted>;
 
     fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        assert!(!self.complete, "It is illegal to poll a completed future");
```
Comment (alamb) [inline tracker/src/task/future.rs:63]: 🏅

File: tracker/src/task.rs [hunk 2]
```diff
@@ -279,11 +332,29 @@ where
...
                 cpu_nanos: self.state.cpu_nanos.load(Ordering::Relaxed),
             },
-            (true, true) => TaskStatus::Complete {
-                total_count: self.state.created_futures.load(Ordering::Relaxed),
-                cpu_nanos: self.state.cpu_nanos.load(Ordering::Relaxed),
-                wall_nanos: self.state.wall_nanos.load(Ordering::Relaxed),
-            },
+            (true, true) => {
+                let total_count = self.state.created_futures.load(Ordering::Relaxed);
+                let success_count = self.state.ok_futures.load(Ordering::Relaxed);
```
Comment (alamb) [inline tracker/src/task.rs:337]: I like this more detailed accounting of errors 👍

File: tracker/src/task.rs [hunk 3]
```diff
@@ -721,13 +804,112 @@ mod tests {
...
+                dropped_count: 0,
+                cpu_nanos: 0,
+                wall_nanos: 0
+            }
+        );
+
+        let (task, registration) = registry.register(());
+        std::mem::drop(pending().track(registration));
+
+        assert_eq!(task.get_status().result(), Some(TaskResult::Dropped));
```
Comment (alamb) [inline tracker/src/task.rs:898]: the documents say "dropped' means a likely panic! but in this case drop seems to result from simply never `.await` ing / spawning the future. 

Suggestions:
1. Add a test with a task that explicitly `panic!`s to make sure it is accounted for correctly
2. update the documents to note `aborted` can also happen without a panic

**Review by alamb — APPROVED** [review]
This looks great to me. Nice work @tustvold  -- improving error reporting always pays off in my opinion.

File: tracker/src/task.rs (hunk 3 above)
Comment (tustvold) [reply tracker/src/task.rs:899]: I added a test, updated the docs, and renamed "aborted" to "cancelled" to avoid confusion

**Pull request merged**